src/metabase/util/malli.clj

```diff
@@ -0,0 +1,154 @@
+    "\n" (spaces ctx) "]"]))
+
+(defmethod describe* :map-of [{:keys [key value] :as ast} ctx]
+  (str "map of ["
+       (str/join (str "\n" (spaces ctx))
+                 [""
+                  (describe* key (indent+ ctx))
+                  "=>"
+                  (str (describe* value (indent+ ctx)) "]")])))
+
+(defmethod describe* :tuple [{:keys [children] :as ast} ctx]
+  (str "tuple of size " (count children) " like: [ " (str/join (map #(describe* % ctx) children)) "]"))
+
+(defn- insert-or [xs]
+  (if (> (count xs) 1)
+    (concat (drop-last xs) ["or"] [(last xs)])
+    xs))
+
+(defmethod describe* :enum [{:keys [values] :as _ast} ctx]
+  (str "value that is one of: "
+       (str/join " " (insert-or (map pr-str values)))))
+
+(core/defn take-measure [title properties]
+  (let [min-len (:min properties)
+        max-len (:max properties)]
+    (cond
+      (and min-len max-len) (str "with " title " between " min-len " and " max-len " ")
+      min-len               (str "at least " min-len " long ")
+      max-len               (str "at most " max-len " long ")
+      :else                 "")))
+
+(core/defn generic-props [properties description]
+  (cond-> description
+    (:title properties)
+    (str " (title: " (:title properties) ")")))
+
+(defmethod describe* :string [{:keys [properties]} _ctx]
+  ;; todo handle min/max/other properties.
+  (str "string " (take-measure "length" properties)))
+(defmethod describe* 'string? [ast ctx] (describe* (assoc ast :type :string) ctx))
+
+(defmethod describe* :keyword [ast _ctx] (generic-props (:properties ast) "keyword "))
+(defmethod describe* 'keyword? [ast ctx] (describe* (assoc ast :type :keyword) ctx))
+
+(defmethod describe* :int [ast _ctx] (generic-props (:properties ast) "integer "))
+(defmethod describe* 'int? [ast ctx] (describe* (assoc ast :type :int) ctx))
+
+(defmethod describe* :pos-int [ast _ctx] (generic-props (:properties ast) "positive integer "))
+(defmethod describe* 'pos-int? [ast ctx] (describe* (assoc ast :type :pos-int) ctx))
+
+(defmethod describe* :nat-int [ast _ctx] (generic-props (:properties ast) "natural integer "))
+(defmethod describe* 'nat-int? [ast ctx] (describe* (assoc ast :type :nat-int) ctx))
+
+(defmethod describe* := [{:keys [value] :as ast} ctx]
+  (str "value that equals " (pr-str value)))
+
+(defmethod describe* :default [x ctx]
+  ;; TODO put custom descriptions _inside_ the malli schemas, and pick it up in here
+  (reset! (:*missing? ctx) true)
+  (str "Indescribable_Malli_Schema [ " (pr-str (or (:type x) x)) " ]"))
+
+;; n.b. this is not clojure.core/defn
+(defn describe :- [:map [:missing? :boolean] [:description :string]]
+  "Given a malli schema, should return a string with a description of the shape it expects."
+  [schema]
+  (let [*missing? (atom false)
+        ast (mc/ast schema)
+        raw-output (describe* ast {:indent 0 :*missing? *missing?})]
+    {:missing? @*missing?
+     :description (str/trim (str "A " raw-output))}))
```

Comment:
I think it'd be cool to include a link to malli.io with the failed schema, either without a value [like this](https://malli.io/?schema=%5B%3Amap%0A%20%20%5B%3Arevision%20int%3F%5D%0A%20%20%5B%3Agroups%0A%20%20%20%5B%3Amap-of%0A%20%20%20%20%20%5Bpos-int%3F%20%7B%3Atitle%20%22Group%20ID%22%7D%5D%0A%20%20%20%20%20%5B%3Amap-of%0A%20%20%20%20%20%20%5B%3Aor%0A%20%20%20%20%20%20%20%20%5Bpos-int%3F%20%7B%3Atitle%20%22Collection%20ID%22%7D%5D%0A%20%20%20%20%20%20%20%20%5B%3Aand%20keyword%3F%20%5B%3A%3D%20%3Aroot%5D%5D%5D%0A%20%20%20%20%20%20%5B%3Aand%20keyword%3F%20%5B%3Aenum%20%3Awrite%20%3Aread%20%3Anone%5D%5D%5D%5D%5D%5D) or with a value [like this](https://malli.io/?schema=%5B%3Amap%0A%20%20%5B%3Arevision%20int%3F%5D%0A%20%20%5B%3Agroups%0A%20%20%20%5B%3Amap-of%0A%20%20%20%20%20%5Bpos-int%3F%20%7B%3Atitle%20%22Group%20ID%22%7D%5D%0A%20%20%20%20%20%5B%3Amap-of%0A%20%20%20%20%20%20%5B%3Aor%0A%20%20%20%20%20%20%20%20%5Bpos-int%3F%20%7B%3Atitle%20%22Collection%20ID%22%7D%5D%0A%20%20%20%20%20%20%20%20%5B%3Aand%20keyword%3F%20%5B%3A%3D%20%3Aroot%5D%5D%5D%0A%20%20%20%20%20%20%5B%3Aand%20keyword%3F%20%5B%3Aenum%20%3Awrite%20%3Aread%20%3Anone%5D%5D%5D%5D%5D%5D&value=%7B%3Arevision%20%223%22%0A%20%3Agroups%20%7B1%20%7B1%20%22read%22%7D%7D%7D)